client/network/src/protocol/generic_proto/handler/notif_out.rs

```diff
@@ -308,16 +312,17 @@ impl ProtocolsHandler for NotifsOutHandler {
 
 			NotifsOutHandlerIn::Send(msg) =>
 				if let State::Open { substream, .. } = &mut self.state {
-					if let Some(Ok(_)) = substream.send(msg).now_or_never() {
-						if let Some(metric) = &self.queue_size_report {
-							metric.observe(substream.queue_len() as f64);
-						}
-					} else {
+					if substream.push_message(msg).is_err() {
 						log::warn!(
 							target: "sub-libp2p",
-							"📞 Failed to push message to queue, dropped it"
+							"📞 Notifications queue with peer {} is full, dropped message (protocol: {:?})",
+							self.peer_id,
+							self.protocol_name,
 						);
 					}
+					if let Some(metric) = &self.queue_size_report {
+						metric.observe(substream.queue_len() as f64);
```

Comment:
In regards to observing even though the message is dropped: If we only `observe` when the message was actually sent, the histogram `count` metric would tell us how many messages went into the channel. Later on we could add an additional metric (a simple counter) to count the messages being dropped. What do you think @tomaka? Would that enable us to deeper understand the behavior?

Reply:
Totally fine if you see this more as a hack than a useful feature.

Reply:
The way I see it, dropping messages is something that is really not supposed to happen, and knowing how many messages are dropped isn't necessarily useful, as we have to treat 1 dropped message the same as 500 dropped messages.

Additionally #5481 would consist in reworking this code anyway.